src/rgw/rgw_dedup_cluster.cc

```diff
@@ -0,0 +1,1187 @@
+#include "rgw_dedup_cluster.h"
+#include "rgw_dedup.h"
+#include "rgw_dedup_epoch.h"
+#include "rgw_common.h"
+#include "rgw_dedup_store.h"
+#include "include/rados/rados_types.hpp"
+#include "include/rados/buffer.h"
+#include "include/rados/librados.hpp"
+#include "svc_zone.h"
+#include "common/config.h"
+#include "common/Cond.h"
+#include "common/debug.h"
+#include "common/errno.h"
+#include "rgw_common.h"
+#include "include/denc.h"
+#include "rgw_sal.h"
+#include "driver/rados/rgw_sal_rados.h"
+#include <cstdlib>
+#include <ctime>
+#include <string>
+
+namespace rgw::dedup {
+  const char* DEDUP_EPOCH_TOKEN = "EPOCH_TOKEN";
+
+  static constexpr unsigned EPOCH_MAX_LOCK_DURATION_SEC = 30;
+  struct shard_progress_t;
+  static int collect_shard_stats(librados::IoCtx &ioctx,
+				 const DoutPrefixProvider *dpp,
+				 utime_t epoch_time,
+				 unsigned shards_count,
+				 const char *prefix,
+				 bufferlist bl_arr[],
+				 struct shard_progress_t *sp_arr);
+
+  const uint64_t SP_ALL_OBJECTS = ULLONG_MAX;
+  const uint64_t SP_NO_OBJECTS  = 0ULL;
+  const char* SHARD_PROGRESS_ATTR = "shard_progress";
+
+  //---------------------------------------------------------------------------
+  static int get_epoch(librados::IoCtx &ioctx,
+		       const DoutPrefixProvider *dpp,
+		       dedup_epoch_t *p_epoch, /* OUT */
+		       const char *caller)
+  {
+    std::string oid(DEDUP_EPOCH_TOKEN);
+    bufferlist bl;
+    int ret = ioctx.getxattr(oid, RGW_DEDUP_ATTR_EPOCH, bl);
+    if (ret > 0) {
+      try {
+	auto p = bl.cbegin();
+	decode(*p_epoch, p);
+      }catch (const buffer::error&) {
+	ldpp_dout(dpp, 0) << __func__ << "::failed epoch decode!" << dendl;
+	return -EINVAL;
+      }
+      if (caller) {
+	ldpp_dout(dpp, 10) << __func__ << "::"<< caller<< "::" << *p_epoch << dendl;
```

Comment:
why do you need `caller`?
`dpp` should give the log message its overall context.